src/System.Linq/src/System/Linq/Concat.cs

```diff
@@ -81,327 +73,64 @@ internal override IEnumerable<TSource> GetEnumerable(int index)
         {
-            private readonly ConcatIterator<TSource> _previousConcat;
-            private readonly IEnumerable<TSource> _next;
-            private readonly int _nextIndex;
-
-            internal ConcatNEnumerableIterator(ConcatIterator<TSource> previousConcat, IEnumerable<TSource> next, int nextIndex)
-            {
-                Debug.Assert(previousConcat != null);
-                Debug.Assert(next != null);
-                Debug.Assert(nextIndex >= 2);
-                _previousConcat = previousConcat;
-                _next = next;
-                _nextIndex = nextIndex;
-            }
-
-            public override Iterator<TSource> Clone()
-            {
-                return new ConcatNEnumerableIterator<TSource>(_previousConcat, _next, _nextIndex);
-            }
+            private readonly SingleLinkedNode<IEnumerable<TSource>> _sources;
+            private readonly int _headIndex;
 
-            internal override ConcatIterator<TSource> Concat(IEnumerable<TSource> next)
+            internal ConcatNIterator(SingleLinkedNode<IEnumerable<TSource>> sources, int headIndex)
             {
-                if (_nextIndex == int.MaxValue - 2)
-                {
-                    // In the unlikely case of this many concatenations, if we produced a ConcatNEnumerableIterator
-                    // with int.MaxValue then state would overflow before it matched its index.
-                    // So we use the naïve approach of just having a left and right sequence.
-                    return new Concat2EnumerableIterator<TSource>(this, next);
-                }
+                Debug.Assert((headIndex & int.MaxValue) >= 2);
+                Debug.Assert(sources?.GetCount() == (headIndex & int.MaxValue) + 1);
 
-                return new ConcatNEnumerableIterator<TSource>(this, next, _nextIndex + 1);
+                _sources = sources;
+                _headIndex = headIndex;
             }
 
-            internal override IEnumerable<TSource> GetEnumerable(int index)
-            {
-                if (index > _nextIndex)
-                {
-                    return null;
-                }
-
-                // Walk back through the chain of ConcatNIterators looking for the one
-                // that has its _nextIndex equal to index.  If we don't find one, then it
-                // must be prior to any of them, so call GetEnumerable on the previous
-                // Concat2Iterator.  This avoids a deep recursive call chain.
-                ConcatNEnumerableIterator<TSource> current = this;
-                while (true)
-                {
-                    if (index == current._nextIndex)
-                    {
-                        return current._next;
-                    }
-
-                    ConcatIterator<TSource> previous = current._previousConcat;
+            internal override bool AllSourcesAreCollections => _headIndex < 0;
```

Comment:
Self-note: I think this can be moved into its own field later, since it shouldn't increase field size.